Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset

bdhirsh opened the pull request:
Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88198
* #88230

bdhirsh committed:
*_scatter ops should preserve input stride/storage_offset

[ghstack-poisoned]

bdhirsh committed:
Update on "*_scatter ops should preserve input stride/storage_offset"




[ghstack-poisoned]

bdhirsh commented:
Darn, I'm getting the cursed error:
```
FAILED test/functorch/test_aotdispatch.py::TestEagerFusionOpInfoCPU::test_aot_autograd_symbolic_exhaustive_nn_functional_feature_alpha_dropout_with_train_cpu_float32
FAILED test/functorch/test_aotdispatch.py::TestEagerFusionOpInfoCPU::test_aot_autograd_symbolic_exhaustive_nn_functional_poisson_nll_loss_cpu_float32
...
RuntimeError: s0 is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f19eca096a0>
```

Maybe, somehow when we regenerate the views for a mutated view, that regenerate tensor isn't getting proxies attached to its sizes/strides

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

ezyang commented:
So I am actually not sure why you need to sync_ the metadata on each pass now. Don't you just need to update the inplace tensor metadata; the other tensors not changing their sizes and strides is fine, because whatever they used to have, should still be valid!

bdhirsh commented:
Yeah, I think you're right. At first I was worried about other views of the base that also came from slice/select calls, but their metadata should be accurate to begin with - the only time the metadata becomes "wrong" is when we convert an inplace op into an out-of-place.

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

ezyang commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -142,13 +142,13 @@ FunctionalTensorWrapper::FunctionalTensorWrapper(const Tensor& view_value, const
   }
   view_metas_.push_back(meta);
   storage_ = base->storage_; // alias this tensor's storage with the base tensor's
+  set_constructor_metadata();
```
not sure why this had to move

bdhirsh replied:
I can move it back - it was necessary before when I was trying to sync() before metadata calls. `set_constructor_metadata()` ended up invoking a size or stride call somewhere, which involved trying to access the storage before we properly set it.

ezyang commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -336,30 +340,38 @@ c10::intrusive_ptr<TensorImpl> FunctionalTensorWrapper::shallow_copy_and_detach(
   return value_.unsafeGetTensorImpl()->sym_size(d);
 }
 c10::SymInt FunctionalTensorWrapper::sym_storage_offset_custom() const {
+  //const_cast<FunctionalTensorWrapper*>(this)->sync_();
```
delete these when you're done plz

bdhirsh replied:
will do

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3757,22 +3757,59 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
+  // If the input to the select_scatter is grad_output, then this could be an expanded tensor
+  // with internal overlap.
+  if (at::has_internal_overlap(self) == at::MemOverlap::Yes) {
+    return self.clone();
```
You sure you don't want to just error here?

bdhirsh replied:
So... maybe I should investigate this more. There was a test failure for the `select_scatter` opinfo `test_aotdispatch` test. The test itself didn't involve any mutations (so functionalization wasn't manually inserting a `select_scatter` anywhere). But the OpInfo calls `select_scatter` in the forward, which caused `select_scatter` to show up in the backward, where in that test case, the input to it (`grad_output`) had overlapping memory, probably because it came from an `expand()` call.

bdhirsh replied:
My thought was that this matters less, because... as long as functionalization runs underneath autograd, then autograd won't see any `*_scatter` ops directly, so we won't have to worry about those ops showing up in the backward. It's only really a problem if user code is using `*_scatter` ops, which they (probably) shouldn't be.

bdhirsh replied:
I guess we could error here. But that would basically make `*_scatter` op calls in user code completely unusable.

I could see that being what we want, anyway? Kind of annoying though, because it's nice to have some op info coverage for the `*_scatter` ops, even if they wouldn't work at all with autograd.

ezyang replied:
OK, the expand thing seems valid. You can keep it as is.

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3757,22 +3757,59 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
+  }
+  auto dtype_size = self.dtype().itemsize();
+  auto nbytes = self.storage().sym_nbytes();
+  TORCH_INTERNAL_ASSERT(nbytes % dtype_size == 0);
```
This is not guaranteed. You can have dtype mismatch between tensor and storage, where storage is complex, and tensor is float, and then the modulus doesn't apply.

Read meta converter for the full algorithm for how to do this.

ezyang replied:
Actually, hmph, you're going to trace through this, maybe that doesn't work...

bdhirsh replied:
Yeah, that seems pretty unfriendly to tracing.

Since we specialize on dtype, maybe I can just special case the complex/non-complex mismatch case? I'll take a look at the meta converter algorithm.

bdhirsh replied:
actually - sorry - I don't think we're going to be tracing through this code?

We **want** `*_scatter` ops to show up as primitives in the trace. The only reason I need to symintify this function is because I need dynamic shape propagation through this function to work, so FakeTensor an handle `*_scatter` ops properly.

In that case, the untyped storage thing sounds right

ezyang replied:
OK, if this is not traced then this is basically fine

ezyang commented on test/test_functionalization.py:
```diff
@@ -1047,19 +990,11 @@ def forward(self, a_1):
-    return as_strided_2
+    view_7 = torch.ops.aten.view.default(view_4, [4, 4]);  view_4 = None
+    as_strided_3 = torch.ops.aten.as_strided.default(view_7, [3, 3], [3, 1]);  view_7 = None
+    return as_strided_3
```
Do you know why the traces got so much shorter? DCE?

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3757,22 +3757,59 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
+  auto nbytes = self.storage().sym_nbytes();
+  TORCH_INTERNAL_ASSERT(nbytes % dtype_size == 0);
+  auto numel = nbytes / dtype_size;
+  auto self_full_size = self.as_strided_symint({numel}, {1}, 0);
```
I feel like this trace will be pretty difficult to compile, since you're taking the input tensor and then seeing "more" of it than was accessible from the original tensor. That's one of the naughty uses of as_strided. It seems better to allocate a zero tensor, scatter the view into it, and then view on that; this prevents the naughty use of as_sttrided.

bdhirsh replied:
Ok that's fair. I thought this was "more" correct, since if there are any "holes" in the tensor then we'll fill it in with the correct data instead of zeros, but I guess that shouldn't really matter - that data never gets used anywhere

bdhirsh replied:
If you agree with https://github.com/pytorch/pytorch/pull/88198#discussion_r1012146315, then i think this point is moot

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -143,24 +144,20 @@ def forward(self, a_1):
-    sum_1 = torch.ops.aten.sum.default(relu)
+    view_copy_2 = torch.ops.aten.view_copy.default(relu, [1, 1024, 128, 128]);  relu = None
+    view_copy_3 = torch.ops.aten.view_copy.default(view_copy_2, [16, 64, 128, 128]);  view_copy_2 = None
+    sum_1 = torch.ops.aten.sum.default(view_copy_3)
```
I'm still auditing the expect test changes.

One annoying source of changes is extra views that DCE won't eliminate automatically, but are technically unnecessary. In this example, `view_copy_3` is completely equivalent to `relu`.

I suppose we could get functionalization to figure out when this happens and "fix" it - do you think it would be reasonable for downstream consumers to optimize these away?

bdhirsh replied:
The other major source of test wobbling is definitely DCE. Some deleted lines, which cause some variable names to wobble in the graph.

ezyang replied:
Views should be optimized away, I think this is fine.

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -345,9 +331,7 @@ def f(x):
-    view_copy = torch.ops.aten.view_copy.default(a_1, [4, 2])
-    add = torch.ops.aten.add.Tensor(a_1, ones);  ones = None
-    copy_ = torch.ops.aten.copy_.default(a_1, add);  a_1 = None
+    add = torch.ops.aten.add.Tensor(a_1, ones);  a_1 = ones = None
```
bleh... I'll update the test so DCE doesn't run in cases where the graph we print is the result of reinplacing (we want to see the mutations in our expecttest output).

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -3757,22 +3757,59 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
+// operators that it calls must preserve the striding behavior of their inputs.
+// Specifically, the output of *_scatter(base, mutated_view, ...)
+// should have identical size/stride/storage_offset to "base".
+at::Tensor clone_preserve_strides(const at::Tensor& self) {
```
nit: maybe you want to use the _new_zeros_with_same_feature_meta function that we created for forward AD (it supports vmap).
We could add a flag to choose between filled with zeros vs clone.

albanD reviewed (approved):
Sounds good. Small nit for the helper function.

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

bdhirsh committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

ezyang committed:
Update on "Update functionalization metadata more eagerly. *_scatter ops should preserve input stride/storage_offset"

Two major changes in this PR:


(1) outputs of `*_scatter` ops will now reflect the same storage size, stride and storage_offset of their inputs. See more details at https://github.com/pytorch/pytorch/pull/87610/files#r1007264456. This fixes some silent correctness issues with functionalization advertising incorrect strides/storage_offsets on some tensors.

(2) That actually isn't enough: We need to ensure that any time someone calls `.stride()` or `.storage_offset()` on a tensor, its metadata is not stale. To fix this, I updated all of the custom metadata calls on `FunctionalTensorWrapper` to perform a sync first if metadata is out of date.

As a motivating example, consider this code:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
a_diag.add_(1)
```

Functionalization will temporarily turn this code into:
```
a = torch.ones(2, 2)
a_diag = torch.diagonal(a, 0, 1)
# a_diag_updated has incorrect metadata! The output of add() is always a contiguous, densely packed tensor.
# But a_diag_updated's metadata should advertise as (properly) not being contiguous! it has different strides
a_diag_updated = a_diag.add(1)
```

As mentioned in the above comment, this isn't 100% correct - `a_diag_updated` has different metadata than `a_diag`. If user code queries the metadata on `a_diag` at this point, we'll return incorrect metadata.

The fix: By eagerly running syncing logic before any metadata on `a_diag` is accessed, we'll do the following:
```
# reapply the mutation to the base, a
a_updated = torch.diagonal_scatter(a, a_diag.add(1))
# regenerate a_diag with proper strides, from a_updated
a_diag_updated = torch.diagonal(a_updated, 0, 1)
```

This ensures that any user code that calls `a_diag.stride()` or `a_diag.storage_offset()` sees accurate metadata





[ghstack-poisoned]

ezyang commented:
There are a bunch of things going on here, but when I ablate everything except the clone preserving strides changes (https://github.com/pytorch/pytorch/pull/89474) it still fails `python test/test_ops_fwd_gradients.py -k test_forward_mode_AD_as_strided_scatter_cpu_complex128`

ezyang commented:
So, I'm a little confused about whether or not this PR is still necessary. The stated justification was to make "do not use unsafe restriding for subclasses" PR be able to go in, c.f. https://github.com/pytorch/pytorch/pull/87610/files/401ddeda1d769bedc88a12de332c7357b60e51a4#diff-c4e35e76279419d9edda979a995412ee494fad3a97d3fb59486a1f0e326d48be But as best as I can tell, this already has gone in to master without trouble. So is this just for a hypothetical situation where functionalization can produce incorrect strides?

ezyang commented:
we think this is obsolete
